test/common/event/dispatcher_impl_test.cc

```diff
@@ -51,5 +54,71 @@ TEST(DispatcherImplTest, DeferredDelete) {
   dispatcher.clearDeferredDeleteList();
 }
 
+class DispatcherImplTest : public ::testing::Test {
+protected:
+  DispatcherImplTest() : dispatcher_(std::make_unique<DispatcherImpl>()), work_finished_(false) {
+    dispatcher_thread_ = std::make_unique<Thread::Thread>([this]() {
+      // Must create a keepalive timer to keep the dispatcher from exiting.
+      std::chrono::milliseconds time_interval(500);
+      keepalive_timer_ = dispatcher_->createTimer(
+          [this, time_interval]() { keepalive_timer_->enableTimer(time_interval); });
+      keepalive_timer_->enableTimer(time_interval);
+
+      dispatcher_->run(Dispatcher::RunType::Block);
+    });
+  }
+
+  ~DispatcherImplTest() override {
+    dispatcher_->exit();
+    dispatcher_thread_->join();
+  }
+
+  std::unique_ptr<Thread::Thread> dispatcher_thread_;
+  DispatcherPtr dispatcher_;
+  std::mutex mu_;
+  std::condition_variable cv_;
+
+  bool work_finished_;
+  TimerPtr keepalive_timer_;
+};
+
+TEST_F(DispatcherImplTest, Post) {
+  dispatcher_->post([this]() {
+    {
+      std::lock_guard<std::mutex> lock(mu_);
+      work_finished_ = true;
+    }
+    cv_.notify_one();
+  });
+
+  std::unique_lock<std::mutex> lock(mu_);
+
+  EXPECT_TRUE(cv_.wait_for(lock, std::chrono::seconds(5), [this]() { return work_finished_; }));
```

Comment:
CI can get absurdly slow and I could see this flaking sadly. My preference would be to not have a timeout and just let it hang if there is an issue, or I would use a much longer sanity timeout like 60s or something. Same below.